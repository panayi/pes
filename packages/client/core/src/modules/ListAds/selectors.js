import * as R from 'ramda';
import { isPlainObj } from 'ramda-adjunct';
import { createSelector } from 'reselect';
import createCachedSelector from 're-reselect';
import propSelector from '@pesposa/core/src/utils/propSelector';
import id from '@pesposa/core/src/utils/id';
import * as constants from './constants';

const hitsSelector = propSelector('hits');
const hitSelector = propSelector('hit');
const sizeSelector = propSelector('size');
const fixedCardHeightSelector = propSelector('fixedCardHeight');
const hitIdSelector = R.compose(id, hitSelector);
const containerWidthSelector = propSelector('containerWidth');
const indexSelector = propSelector('index');
const columnHeightsSelector = propSelector('columnHeights');

const currentMinWidthSelector = createSelector(
  containerWidthSelector,
  sizeSelector,
  (containerWidth, size) =>
    R.find(R.gte(containerWidth), constants.MIN_WIDTHS(size)),
);

export const columnCountSelector = createSelector(
  sizeSelector,
  currentMinWidthSelector,
  R.compose(R.path(R.__, constants.COLUMN_COUNTS), R.unapply(R.identity)),
);

// columnWidthSelector :: { containerWidth } -> Number
export const columnWidthSelector = createSelector(
  containerWidthSelector,
  columnCountSelector,
  (containerWidth, columnCount) =>
    Math.max(
      0,
      (containerWidth + constants.GUTTER) / columnCount - constants.GUTTER,
    ),
);

// thumbnailSelector :: { hit } -> Thumbnail
//  Thumbnail = { url, height, dimensions }
export const thumbnailSelector = createCachedSelector(
  hitSelector,
  propSelector(['thumbnail']),
)(hitIdSelector);

export const thumbnailHeightSelector = createCachedSelector(
  fixedCardHeightSelector,
  thumbnailSelector,
  propSelector('columnWidth'),
  (fixedCardHeight, thumbnail, columnWidth) => {
    if (fixedCardHeight) {
      return fixedCardHeight - constants.CARD_CONTENT_HEIGHT;
    }

    const dimensions = R.prop('dimensions', thumbnail);
    return isPlainObj(dimensions)
      ? columnWidth / dimensions.width * dimensions.height
      : constants.DEFAULT_THUMBNAIL_HEIGHT;
  },
)(hitIdSelector);

// hitHeightSelector :: { hit, columnWidth } => Number
const hitHeightSelector = ({ hit, columnWidth }) => {
  if (hit.isPaidAd) {
    return hit.height + constants.PAID_AD_VERTICAL_PADDING * 2;
  }

  const thumbnailHeight = thumbnailHeightSelector({ hit, columnWidth });
  return Math.ceil(thumbnailHeight + constants.CARD_CONTENT_HEIGHT);
};

export const hitSizeAndPositionSelector = createSelector(
  indexSelector,
  hitsSelector,
  columnHeightsSelector,
  columnCountSelector,
  columnWidthSelector,
  fixedCardHeightSelector,
  (index, hits, columnHeights, columnCount, columnWidth, fixedCardHeight) => {
    const hit = hits[index];
    const minColumnHeight = R.reduce(
      R.min,
      Number.POSITIVE_INFINITY,
      columnHeights,
    );
    const column = R.findIndex(R.equals(minColumnHeight), columnHeights);
    const height =
      fixedCardHeight ||
      (hit && hitHeightSelector({ hit, columnWidth })) ||
      constants.DEFAULT_CARD_HEIGHT;
    const width = columnWidth;
    const x = column * (constants.GUTTER + width);
    const y = columnHeights[column] || 0;

    return {
      column,
      height,
      width,
      x,
      y,
    };
  },
);
