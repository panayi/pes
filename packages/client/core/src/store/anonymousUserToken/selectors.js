import * as R from 'ramda';
import * as constants from './constants';

// anonymousUserTokenSelector :: State -> Object | Nil
export const anonymousUserTokenSelector = R.prop(constants.ROOT_KEY);
