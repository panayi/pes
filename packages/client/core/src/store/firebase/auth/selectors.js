import * as R from 'ramda';
import { isNotNil } from 'ramda-adjunct';
import { createSelector } from 'reselect';
import { isLoaded } from 'react-redux-firebase';
import * as firebaseConfig from '@pesposa/core/src/config/firebase';
import decodeJwt from '@pesposa/core/src/utils/decodeJwt';

const FIREBASE_AUTH_PATH = [...firebaseConfig.FIREBASE_PATH, 'auth'];

// authErrorSelector :: State -> Object | Nil
export const authErrorSelector = R.path([
  ...firebaseConfig.FIREBASE_PATH,
  'authError',
]);

// firebaseSelector :: State -> Object
const firebaseAuthSelector = R.compose(
  R.defaultTo({}),
  R.path(FIREBASE_AUTH_PATH),
);

// isLoadedSelector :: State -> Boolean | Nil
const isLoadedSelector = createSelector(firebaseAuthSelector, isLoaded);

// isInitializingSelector :: State -> Boolean | Nil
const isInitializingSelector = R.path([
  ...firebaseConfig.FIREBASE_PATH,
  'isInitializing',
]);

export const uidSelector = createSelector(firebaseAuthSelector, R.prop('uid'));

// hasUidSelector :: State -> String | Nil
export const hasUidSelector = createSelector(uidSelector, isNotNil);

// isAnonymousSelector :: State -> Boolean | Nil
export const isAnonymousSelector = createSelector(
  firebaseAuthSelector,
  R.prop('isAnonymous'),
);

// isAuthenticatingSelector :: State -> Boolean
export const isAuthenticatingSelector = createSelector(
  isInitializingSelector,
  isLoadedSelector,
  R.useWith(R.or, [R.identity, R.not]),
);

// isAuthenticatedSelector :: State -> Boolean
export const isAuthenticatedSelector = createSelector(
  isAuthenticatingSelector,
  hasUidSelector,
  isAnonymousSelector,
  (isAuthenticating, hasUid, isAnonymous) =>
    !isAuthenticating && hasUid && !isAnonymous,
);

// isNotAuthenticatedSelector :: State -> Boolean
// Equals to: !isAuthenticatingSelector && !hasUidSelector
export const isNotAuthenticatedSelector = createSelector(
  isAuthenticatingSelector,
  hasUidSelector,
  isAnonymousSelector,
  (isAuthenticating, hasUid, isAnonymous) =>
    !isAuthenticating && (!hasUid || isAnonymous),
);

export const tokenSelector = createSelector(
  firebaseAuthSelector,
  R.path(['stsTokenManager', 'accessToken']),
);

export const isCurrentUserSelector = userIdSelector =>
  createSelector(
    isAuthenticatedSelector,
    uidSelector,
    userIdSelector,
    (isAuthenticated, currentUserId, userId) =>
      isAuthenticated && R.equals(currentUserId, userId),
  );

export const isAdminSelector = createSelector(
  tokenSelector,
  R.compose(R.prop('admin'), R.defaultTo({}), decodeJwt),
);
