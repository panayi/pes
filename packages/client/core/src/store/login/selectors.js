import * as R from 'ramda';
import { createSelector } from 'reselect';
import * as constants from './constants';

export const statusSelector = R.path(constants.STATUS_PATH);

const createFlagSelector = value =>
  createSelector(statusSelector, R.equals(value));

export const pendingSelector = createFlagSelector(
  constants.LOGIN_PENDING_STATUS,
);
export const succeededSelector = createFlagSelector(
  constants.LOGIN_SUCCESS_STATUS,
);
export const failedSelector = createFlagSelector(constants.LOGIN_FAIL_STATUS);
