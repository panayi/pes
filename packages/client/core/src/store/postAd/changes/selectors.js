import * as R from 'ramda';
import * as constants from './constants';

export const createdAdsSelector = R.path(constants.CREATED_ADS_PATH);

export const deletedAdsSelector = R.path(constants.DELETED_ADS_PATH);
