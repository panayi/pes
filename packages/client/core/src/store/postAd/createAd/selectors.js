import * as R from 'ramda';
import { createSelector } from 'reselect';
import * as constants from './constants';

const createAdStatusSelector = R.path(constants.CREATE_AD_STATUS_PATH);
export const createAdErrorSelector = R.path(constants.CREATE_AD_ERROR_PATH);

export const isCreateAdIdleSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_IDLE_STATE),
);

export const isCreateAdPendingSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_PENDING_STATE),
);

export const isCreateAdFailedSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_FAILED_STATE),
);

export const isCreateAdCompletedSelector = createSelector(
  createAdStatusSelector,
  R.equals(constants.AD_CREATE_COMPLETED_STATE),
);
