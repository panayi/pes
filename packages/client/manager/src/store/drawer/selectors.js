import * as R from 'ramda';

const drawerPath = ['drawer'];

export const isOpenedSelector = R.path(drawerPath);
