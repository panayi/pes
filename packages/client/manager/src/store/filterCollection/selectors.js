import * as R from 'ramda';
import { isNilOrEmpty, ensureArray } from 'ramda-adjunct';
import createCachedSelector from 're-reselect';
import match from 'autosuggest-highlight/match';
import propSelector from '@pesposa/core/src/utils/propSelector';
import propOrSelector from '@pesposa/core/src/utils/propOrSelector';

const rootPath = ['filterCollection'];
const selectedPath = [...rootPath, 'selected'];
const queryPath = [...rootPath, 'query'];

export const selectedSelector = (state, props) =>
  R.path([...selectedPath, props.id], state);

// queryValueSelector :: (State, Props) -> String
export const queryValueSelector = (state, props) =>
  R.pathOr('', [...queryPath, props.id], state);

const collectionSelector = createCachedSelector(
  propSelector('collection'),
  R.compose(R.filter(R.identity), R.defaultTo([])),
)(propSelector('id'));

// hitsSelector :: (State, Props) => Hits
//   Props: { searchableAttributes: [String], collection: [Object] }
//   Hits: [Object]
export const hitsSelector = createCachedSelector(
  queryValueSelector,
  propOrSelector([], 'searchableAttributes'),
  collectionSelector,
  (query, searchableAttributes, collection) => {
    if (isNilOrEmpty(searchableAttributes) || isNilOrEmpty(collection)) {
      return [];
    }

    if (R.isEmpty(query)) {
      return collection;
    }

    return collection
      .map(item => {
        let matches = {};

        searchableAttributes.forEach(path => {
          const finalPath = ensureArray(path);
          const value = R.path(finalPath, item);
          const result = match(value, query);

          if (!R.isEmpty(result)) {
            matches = R.assocPath(finalPath, result, matches);
          }
        });

        return {
          ...item,
          matches,
        };
      })
      .filter(item => !R.isEmpty(item.matches));
  },
)(queryValueSelector);
