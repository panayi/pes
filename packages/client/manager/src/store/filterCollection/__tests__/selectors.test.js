import * as selectors from '../selectors';

describe('filterCollection', () => {
  describe('selectors', () => {
    describe('hitsSelector', () => {
      const getState = (queryValue = 'foo') => ({
        filterCollection: {
          query: {
            myId: queryValue,
          },
        },
      });
      const searchableAttributes = ['name'];
      const collection = [
        { id: 1, name: 'John' },
        { id: 2, name: 'George' },
        { id: 3, name: 'Christina' },
        { id: 4, name: 'Chris' },
      ];

      it('returns an empty array when searchableAttributes is nil or empty', () => {
        expect(
          selectors.hitsSelector(getState(), {
            searchableAttributes: null,
            collection,
            id: 'myId',
          }),
        ).toEqual([]);
        expect(
          selectors.hitsSelector(getState(), {
            searchableAttributes: [],
            collection,
            id: 'myId',
          }),
        ).toEqual([]);
      });

      it('returns an empty array when collection is nil or empty', () => {
        expect(
          selectors.hitsSelector(getState(), {
            searchableAttributes,
            collection: null,
            id: 'myId',
          }),
        ).toEqual([]);
        expect(
          selectors.hitsSelector(getState(), {
            searchableAttributes,
            collection: [],
            id: 'myId',
          }),
        ).toEqual([]);
      });

      it('returns collection when query is empty', () => {
        expect(
          selectors.hitsSelector(getState(null), {
            searchableAttributes,
            collection,
            id: 'myId',
          }),
        ).toEqual(collection);
        expect(
          selectors.hitsSelector(getState(''), {
            searchableAttributes,
            collection,
            id: 'myId',
          }),
        ).toEqual(collection);
      });

      it('returns items that match query', () => {
        expect(
          selectors.hitsSelector(getState('Chris'), {
            searchableAttributes,
            collection,
            id: 'myId',
          }),
        ).toMatchSnapshot();
      });
    });
  });
});
