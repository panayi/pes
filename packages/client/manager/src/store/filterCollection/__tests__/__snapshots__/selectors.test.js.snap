// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filterCollection selectors hitsSelector returns items that match query 1`] = `
Array [
  Object {
    "id": 3,
    "matches": Object {
      "name": Array [
        Array [
          0,
          5,
        ],
      ],
    },
    "name": "Christina",
  },
  Object {
    "id": 4,
    "matches": Object {
      "name": Array [
        Array [
          0,
          5,
        ],
      ],
    },
    "name": "Chris",
  },
]
`;
